### PR TITLE
feat: support Q4K

### LINKED ISSUE #80: Support 'k-quants' for Q4

### PR DESCRIPTION
**Pull request by guojidan**
closed #80 

I am not sure this pr implement is correct, please carefully review. thanks 😄

### CODE REVIEW

File: crabml-core/src/backends/cpu/buf/api.rs [hunk 1]
```diff
@@ -7,6 +7,7 @@ use super::buf_f16::f16_buf_from_bytes;
 use super::buf_f16::quantize_f32_f16;
 use super::buf_f32::f32_buf_from_bytes;
 use super::buf_f32::vec_dot_f32_f32;
+use super::QuantBufQ4K;
```
Comment (flaneur2020) [inline crabml-core/src/backends/cpu/buf/api.rs:10]: this import could be keep same style like the following lines, like `use crate::backends::cpu::buf::QuantBufQ4K;` : )

File: crabml-core/src/backends/cpu/buf/util.rs [hunk 2]
```diff
@@ -15,6 +15,15 @@ pub fn nearest_i32(fval: f32) -> i32 {
     (i & 0x007fffff) - 0x00400000
 }
 
+#[inline]
+pub fn nearest_int(fval: f32) -> isize {
```
Comment (flaneur2020) [inline crabml-core/src/backends/cpu/buf/util.rs:19]: is this function exactly same as the above one? 

IMHO we could use just reuse the above one

File: crabml-core/src/backends/cpu/buf/util.rs [hunk 3]
```diff
@@ -269,4 +289,78 @@ pub(crate) mod tests {
...
+//         }
+//         uint8_t * q = y[i].qs;
+//         for (int j = 0; j < QK_K; j += 64) {
+//             for (int l = 0; l < 32; ++l) *q++ = L[j + l] | (L[j + l + 32] << 4);
+//         }
+
+//         x += QK_K;
+
+//     }
+// }
```
Comment (flaneur2020) [inline crabml-core/src/backends/cpu/buf/util.rs:366]: it's ok to just delete the commented code here : )

**Comment by guojidan** [comment]
oh my god....

**Comment by guojidan** [comment]
I open a new PR #117

**Pull request closed without merging**